# 第一章

#### 简单生成一个行向量

In [5]:
import numpy as np

In [6]:
a = np.array([1,2,3,4])
a

array([1, 2, 3, 4])

#### 行向量的转置 (无效做法)
    
__注意：__ `.transpose` 方法对一维数组是无效的。

In [8]:
a.transpose()

array([1, 2, 3, 4])

#### 行向量的转置

In [10]:
a_t = a[:,np.newaxis]
a_t

array([[1],
       [2],
       [3],
       [4]])

#### 直接生成一个列向量 (二维数组)

可以把列向量看成是一个 `n * 1` 的矩阵

In [12]:
A = np.array([[1,2,3,4]])
A.T

array([[1],
       [2],
       [3],
       [4]])

#### 向量的内积

可以看成是 u 在 v 上的投影长度乘以 v 的模长，如果 v 为单位向量，则 u 和 v 的内积即为 u 在 v 上的模长。

In [39]:
u = np.array([3,5,2])
v = np.array([1,4,7])
np.dot(u,v)

37

顺序不改变结果

In [50]:
u = np.array([3,5,2])
v = np.array([1,4,7])
np.dot(v,u)

37

__注意__ ：当使用矩阵进行 `.dot` 计算时要满足矩阵乘法的条件。 

可以发现： `a · b = abT`  ，即：`a和b内积` 的结果相当于 `a和b的转置相乘` 的结果

In [48]:
u = np.array([[3,5,2]])
v = np.array([[1,4,7]]).T
np.dot(u,v)

array([[37]])

__注意__： 通常情况下，线性代数下的向量是列向量 即 `n * 1` 的向量。

所以对于两个向量都是 `列向量` 之间的点积为：`a · b = aTb` 

In [57]:
u_t = np.array([[3,5,2]]).T
v_t = np.array([[1,4,7]]).T
np.dot(u_t.T,v_t)

array([[37]])

#### 向量的外积

二维平面中，向量的外积表示两个向量张成的平行四边形的 "面积"

In [15]:
u = np.array([3,5])
v = np.array([1,4])
np.cross(u,v)

array(7)

三维空间中，u 和 v 向量的外积表示两个向量张成平面的法向量

In [16]:
u = np.array([3,3,9])
v = np.array([1,4,12])
np.cross(u,v)

array([  0, -27,   9])

#### 生成一个零矩阵

In [18]:
A = np.zeros([5,3])
A

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

####  生成一个对角矩阵

In [21]:
A = np.diag([1,2,3,4,5])
A

array([[1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [0, 0, 3, 0, 0],
       [0, 0, 0, 4, 0],
       [0, 0, 0, 0, 5]])

####  生成一个单位矩阵

In [20]:
A = np.eye(5)
A

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

#### 矩阵和矩阵相乘

也是使用 `.dot` 方法

In [23]:
A = np.array([[1,2],[3,4],[5,6],[7,8]])
B = np.array([[1,2,3,4],[5,6,7,8]])
np.dot(A,B)

array([[11, 14, 17, 20],
       [23, 30, 37, 44],
       [35, 46, 57, 68],
       [47, 62, 77, 92]])

# 第二章

#### 利用Python求矩阵的秩

In [60]:
A_1 = np.array([[1,1,0],[1,0,1]])y

A_2 = np.array([[1,2,-1],[2,4,-2]])
                
display(np.linalg.matrix_rank(A_1))
display(np.linalg.matrix_rank(A_2))

2

1

#### 利用Python求矩阵的逆

逆矩阵的存在条件：
    
* 1、必须是方阵
    
* 2、矩阵的零空间的维数为0，或者列空间的维数为n。
    
* 3、列向量 a1,a2,a3 ..... an 满足线性无关

In [62]:
from scipy import linalg

In [66]:
A = np.array([[1,35,0],[0,2,3],[0,0,4]])
A_n = linalg.inv(A)
A_n

array([[  1.   , -17.5  ,  13.125],
       [  0.   ,   0.5  ,  -0.375],
       [  0.   ,   0.   ,   0.25 ]])

In [68]:
np.dot(A,A_n)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

#### 利用Python语言求解线性方程

三个方程组求解x：

    x1 + 2x2 + 3x3 = 14
    x1 - x2  + 4x3 = 11
    2x1+ 3x2 - x3  = 5

In [71]:
from scipy import linalg

In [72]:
A = np.array([[1,2,3],[1,-1,4],[2,3,-1]])
y = np.array([14,11,5])
x = linalg.solve(A,y)
x

array([1., 2., 3.])

# 第三章

In [75]:
from scipy import linalg

In [79]:
A = np.array([[2,1],[1,2]])
y = np.array([3,3])
x = linalg.solve(A,y)
x

array([0.66666667, 1.66666667])

#### 当方程无解时，使用投影来找到最佳近似解

    2x + y  = 4
    x  + 2y = 3
    x  + 4y = 9

In [80]:
from scipy import linalg

In [92]:
A = np.array([[2,1],[1,2],[1,4]])
b = np.array([[4],[3],[9]])
A_T_A =np.dot(A.T,A)
x = np.dot(np.dot(linalg.inv(A_T_A),A.T),b)
print('x 和 y 的近似解向量为：\n\n',x)

x 和 y 的近似解向量为：

 [[0.83870968]
 [1.87096774]]


由 p  = Ax 可得

In [91]:
p = np.dot(A,x)
p

array([[3.5483871 ],
       [4.58064516],
       [8.32258065]])

我们称 x = 0.84 , y = 1.87 为线性方程的近似解，正是因为由此得到的

    三维向量  
    [[3.5483871 ],
    [4.58064516],
    [8.32258065]] 

是列空间中距离 原向量

    b = [[4],
        [3],
        [9]] 

距离最近的一个向量。

## 第四章

In [1]:
import numpy as np
from scipy import linalg

求解 `特征值` 和 `特征向量`

In [2]:
A = np.array([[2,1],
             [1,2]])
evalue,evector = linalg.eig(A)
print(evalue)
print(evector)

[3.+0.j 1.+0.j]
[[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]


程序返回的 `特征值` 是用变量 `evalue` 来表示的，分别是 `3` 和 `1`。而变量 `evector` 所表示的是由 `特征向量` 组成的 `特征矩阵`，在这个矩阵中，每一列都是与 `特征值` 依序对应的 `特征向量`。因此 `特征值 3` 所对应的向量为 `[0.7071,0.7071]` （列向量） ,而 `特征值1` 所对应的 `特征向量` 为`[-0.7071,0.7071]` （列向量）。

#### 特征值 `相同` 的情况

可以发现下面 `特征值` 存在相同的情况：

我们发现存在一个 `二重特征值 5` ，和另外一个 `特征值 -2`,虽然特征值相同，但是得到的 `特征向量` 却 `线性无关`

In [3]:
A = np.array([[1,6,0],
              [2,2,0],
              [0,0,5]])
evalue,evector = linalg.eig(A)
print(evalue)
print(evector)

[-2.+0.j  5.+0.j  5.+0.j]
[[-0.89442719 -0.83205029  0.        ]
 [ 0.4472136  -0.5547002   0.        ]
 [ 0.          0.          1.        ]]


而 下面这个 矩阵 A 却没有那么幸运了，可以看出，存在 `二重特征值6`，和一个 `特征值4`，二重特征值6，所对应的 `特征向量` 存在 `线性相关` 的情况。

#### 此时由 `特征向量` 组合而成的 `特征矩阵` 是一个 `不可逆矩阵`，对于 `P` 为 `不可逆矩阵`，矩阵 A 就无法被 `对角化`。

In [5]:
A = np.array([[6,-2,1],
              [0,4,0],
              [0,0,6]])
evalue,evector = linalg.eig(A)
print(evalue)
print(evector)

[6.+0.j 4.+0.j 6.+0.j]
[[ 1.00000000e+00  7.07106781e-01 -1.00000000e+00]
 [ 0.00000000e+00  7.07106781e-01  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  1.33226763e-15]]


__综上：__ 对于一个 `n阶方阵A` ，包括多重特征值一共有`n个特征值`，并且该`矩阵A`一共有`n个线性无关的特征向量`，那么 __由矩阵A的 `特征向量` 组成的 `特征矩阵` 就是 `可逆矩阵` ，`矩阵A可被对角化`__。

### 第五章

#### 一、使用 python 实现简单的 PCA 降维

In [1]:
import numpy as np

In [24]:
x =  [2,2,4,8,4]
y =  [2,6,6,8,8]

S = np.vstack((x,y))

# print(np.vstack((x,y)),'\n','\n',np.vstack((x,y)).shape,'\n')
print(np.vstack((x,y)),'\n')
print(np.cov(S))     # 对角斜上为方差，其他部分为协方差
# print(np.cov(x))     # 方差的无偏估计

[[2 2 4 8 4]
 [2 6 6 8 8]] 

[[6. 4.]
 [4. 6.]]


#### 1. 零均值化数据处理

In [26]:
X = x - np.mean(x)
Y = y - np.mean(y)

S2 = np.vstack((X,Y))
print('中心化后均值为：',np.mean(S2),'\n')
print('中心化后协方差为：\n',np.cov(S2))     # 中心化后 不影响方差和协方差的值

中心化后均值为： 0.0 

中心化后协方差为：
 [[6. 4.]
 [4. 6.]]


#### 2. 通过协方差矩阵C 求得特征向量 

去中心化之后，求解协方差矩阵，相当于: $$C = AA^T$$ 所得到的 C 为对称矩阵，所得到的对称矩阵 C 是一个正定的满秩矩阵，对于对称矩阵还有一个性质：$$S = QλQ^{-1} = QλQ^T  $$

In [27]:
import numpy as np
from scipy import linalg

In [28]:
C = np.array([[6,4],[4,6]])
evalue,evector = linalg.eig(C)
print(evalue)
print(evector)

[10.+0.j  2.+0.j]
[[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]


最后结果竖下来看：

![查看方式](./static/1.jpg)

现在我们得到了两个线性无关的投影正交基，结合新得到的两个线性无关正交基，之后将矩阵在正交基上进行投影：

In [59]:
x =  [2,2,4,8,4]
y =  [2,6,6,8,8]

X = x - np.mean(x)
Y = y - np.mean(y)

A = np.vstack((X,Y))

p_1 = [0.707,0.707]
p_2 = [-0.707,0.707]

P = np.vstack((p_1,p_2))
print(A)
print(np.dot(P,A))

[[-2. -2.  0.  4.  0.]
 [-4.  0.  0.  2.  2.]]
[[-4.242 -1.414  0.     4.242  1.414]
 [-1.414  1.414  0.    -1.414  1.414]]


上面得到两个结果，就是 __原始特征__ 和 __新构建特征__ 分别得到的值，再此基础上再进行特征的 __降维__ ，让2维变成1维，由于两个新特征彼此无关，可以放心的保留一个去掉另外一个。

__那么如何保留呢？__ 我们可以通过 __方差大小__ 来判断，__方差越大__，表示数据越__离散__，那么特征所包含的__信息量越大__，反之越小，我们将保留信息量大的(方差大的)。



#### 3、查看方差大小

$$ D = 1/{(n+1)}Q^TCQ $$

上面求得的 P 就是 Q.T

In [65]:
np.set_printoptions(suppress=True) # 输出时不使用科学计数法

In [66]:
1/4 * np.matmul((np.matmul(P,C)),P.T)

array([[ 2.499245, -0.      ],
       [-0.      ,  0.499849]])

看对角线的值，我们能发现 p_1 的方差为 2.5， p_2 为0.5，保留p_1的特征取值

#### 4、衡量信息损失量

公式：$$ λ_1/ λ_2+λ_3$$

In [68]:
'数据压缩率为50%，用方差来衡量主成分贡献率，贡献率为：{}'.format(2.5/(2.5+0.5))

'数据压缩率为50%，用方差来衡量主成分贡献率，贡献率为：0.8333333333333334'